train.py

```diff
@@ -22,88 +22,182 @@ def _avoid_overwriting(experiment_id):
+        self.dtype = torch.__getattribute__(params["networks"]["dtype"])
+
+        # Load hgn from parameters to deice
+        self.hgn = load_hgn(params=self.params, device=self.device, dtype=self.dtype)
+
+        # Either generate data on-the-fly or load the data from disk
+        if "environment" in self.params:
+            print("Training with ONLINE data...")
+            self.train_data_loader, self.test_data_loader = get_online_dataloaders(self.params)
+        else:
+            print("Training with OFFLINE data...")
+            self.train_data_loader, self.test_data_loader = get_offline_dataloaders(self.params)
+
+        # Initialize training logger
+        self.training_logger = TrainingLogger(hyper_params=self.params,
+                                         loss_freq=1,
+                                         rollout_freq=10,
+                                         model_freq=10000)
+
+        # Initialize tensorboard writer
+        self.model_save_file = os.path.join(self.params["model_save_dir"], self.params["experiment_id"])
+
+        # Define optimization modules
+        optim_params = [
+            {
+                'params': self.hgn.encoder.parameters(),
+                'lr': params["optimization"]["encoder_lr"]
+            },
+            {
+                'params': self.hgn.transformer.parameters(),
+                'lr': params["optimization"]["transformer_lr"]
+            },
+            {
+                'params': self.hgn.hnn.parameters(),
+                'lr': params["optimization"]["hnn_lr"]
+            },
+            {
+                'params': self.hgn.decoder.parameters(),
+                'lr': params["optimization"]["decoder_lr"]
+            },
+        ]
+        self.optimizer = torch.optim.Adam(optim_params)
+
+    def training_step(self, rollouts):
+        """Perform a training step with the given rollouts batch.
+
+        TODO: Move the whole fit() code inside forward if adding hooks to the training, as direct
+            calls to self.forward() do not trigger them.
+
+        Args:
+            rollouts (torch.Tensor): Tensor of shape (batch_size, seq_len, channels, height, width)
+                corresponding to a batch of sampled rollouts.
+            variational (bool): Whether to sample from the encoder and compute the KL loss,
+                or train in fully deterministic mode.
+
+        Returns:
+            A dictionary of losses and the model's prediction of the rollout. The reconstruction loss and
+            KL divergence are floats and prediction is the HGNResult object with data of the forward pass.
+        """
+
+        hgn_output = self.hgn.forward(rollout_batch=rollouts)
+        target = hgn_output.input
+        prediction = hgn_output.reconstructed_rollout
+        
+        if self.params["networks"]["variational"]:    
+            C, rec_loss = geco_constraint(target, prediction, self.params["geco"]["tol"])
```

Comment:
Please, try to put more meaningful names to variables. `C` alone is hard to know what it is